File: agents/addon/extension/elevenlabs_tts/elevenlabs_tts_extension.go
/**
 *
 * Agora Real Time Engagement
 * Created by XinHui Li in 2024-07.
 * Copyright (c) 2024 Agora IO. All rights reserved.
 *
 */
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package extension

import (
	"fmt"
	"io"
	"log/slog"
	"sync"

	"agora.io/rte/rtego"
)

const (
	cmdInFlush                 = "flush"
	cmdOutFlush                = "flush"
	dataInTextDataPropertyText = "text"

	propertyApiKey                = "api_key"                 // Required
	propertyModelId               = "model_id"                // Optional
	propertyRequestTimeoutSeconds = "request_timeout_seconds" // Optional
	propertyVoiceId               = "voice_id"                // Optional
)

const (
	textChanMax = 1024
)

var (
	logTag = slog.String("extension", "ELEVENLABS_TTS_EXTENSION")

	textChan chan string
	wg       sync.WaitGroup
)

type elevenlabsTTSExtension struct {
	rtego.DefaultExtension
	elevenlabsTTS *elevenlabsTTS
}

func newElevenlabsTTSExtension(name string) rtego.Extension {
	return &elevenlabsTTSExtension{}
}

// OnStart will be called when the extension is starting,
// properies can be read here to initialize and start the extension.
// current supported properties:
//   - api_key (required)
//   - model_id
//   - request_timeout_seconds
//   - voice_id
func (e *elevenlabsTTSExtension) OnStart(rte rtego.Rte) {
	slog.Info("OnStart", logTag)

	// prepare configuration
	elevenlabsTTSConfig := defaultElevenlabsTTSConfig()

	if apiKey, err := rte.GetPropertyString(propertyApiKey); err != nil {
		slog.Error(fmt.Sprintf("GetProperty required %s failed, err: %v", propertyApiKey, err), logTag)
		return
	} else {
		elevenlabsTTSConfig.ApiKey = apiKey
	}

	if modelId, err := rte.GetPropertyString(propertyModelId); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyModelId, err), logTag)
	} else {
		if len(modelId) > 0 {
			elevenlabsTTSConfig.ModelId = modelId
		}
	}

	if requestTimeoutSeconds, err := rte.GetPropertyInt64(propertyRequestTimeoutSeconds); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyRequestTimeoutSeconds, err), logTag)
	} else {
		if requestTimeoutSeconds > 0 {
			elevenlabsTTSConfig.RequestTimeoutSeconds = int(requestTimeoutSeconds)
		}
	}

	if voiceId, err := rte.GetPropertyString(propertyVoiceId); err != nil {
		slog.Warn(fmt.Sprintf("GetProperty optional %s failed, err: %v", propertyVoiceId, err), logTag)
	} else {
		if len(voiceId) > 0 {
			elevenlabsTTSConfig.VoiceId = voiceId
		}
	}

	// create elevenlabsTTS instance
	elevenlabsTTS, err := newElevenlabsTTS(elevenlabsTTSConfig)
	if err != nil {
		slog.Error(fmt.Sprintf("newElevenlabsTTS failed, err: %v", err), logTag)
		return
	}

	slog.Info(fmt.Sprintf("newElevenlabsTTS succeed with ModelId: %s, VoiceId: %s",
		elevenlabsTTSConfig.ModelId, elevenlabsTTSConfig.VoiceId), logTag)

	// set elevenlabsTTS instance
	e.elevenlabsTTS = elevenlabsTTS

	// pcm parameters
	sampleRate := int32(16000)
	samplesPer10ms := sampleRate / 100
	channel := int32(1)
	bytesPerSample := int32(2)
	pcmFrameSize := int(samplesPer10ms * channel * bytesPerSample) // per 10ms

	// init chan
	textChan = make(chan string, textChanMax)

	go func() {
		defer func() {
			slog.Info("process textChan quit", logTag)
		}()

		slog.Info("process textChan start", logTag)

		for text := range textChan {
			wg.Add(1)
			slog.Info(fmt.Sprintf("textChan text: [%s]", text), logTag)

			r, w := io.Pipe()

			go func() {
				defer wg.Done()
				defer w.Close()

				slog.Info(fmt.Sprintf("textToSpeechStream text: [%s]", text), logTag)

				err = e.elevenlabsTTS.textToSpeechStream(w, text)
				if err != nil {
					slog.Error(fmt.Sprintf("textToSpeechStream failed, err: %v", err), logTag)
					return
				}
			}()

			slog.Info(fmt.Sprintf("read pcm stream, text: [%s]", text), logTag)

			defer r.Close()

			var n int
			buf := make([]byte, pcmFrameSize)
			for {
				n, err = r.Read(buf)

				slog.Info(fmt.Sprintf("read pcm stream, err: %v, n:%d, pcmFrameSize:%d", err, n, pcmFrameSize), logTag)

				if err == io.EOF {
					break
				}
				if err != nil {
					slog.Error(fmt.Sprintf("read pcm stream failed, err: %v", err), logTag)
					break
				}

				if n != pcmFrameSize {
					slog.Info(fmt.Sprintf("read pcm stream continue, n:%d", n), logTag)
					continue
				}

				pcmFrame, err := rtego.NewPcmFrame("pcm_frame")
				if err != nil {
					slog.Error(fmt.Sprintf("NewPcmFrame failed, err: %v", err), logTag)
					break
				}

				pcmFrame.SetBytesPerSample(bytesPerSample)
				pcmFrame.SetSampleRate(sampleRate)
				pcmFrame.SetChannelLayout(1)
				pcmFrame.SetNumberOfChannels(channel)
				pcmFrame.SetTimestamp(0)
				pcmFrame.SetDataFmt(rtego.PcmFrameDataFmtInterleave)
				pcmFrame.SetSamplesPerChannel(samplesPer10ms)
				pcmFrame.AllocBuf(pcmFrameSize)

				borrowedBuf, err := pcmFrame.BorrowBuf()
				if err != nil {
					slog.Error(fmt.Sprintf("BorrowBuf failed, err: %v", err), logTag)
					break
				}

				// copy data
				copy(borrowedBuf, buf)

				pcmFrame.GiveBackBuf(&borrowedBuf)
				rte.SendPcmFrame(pcmFrame)
				buf = make([]byte, pcmFrameSize)

				slog.Info(fmt.Sprintf("sending pcm data, text: [%s]", text), logTag)
			}

			wg.Wait()
			slog.Info(fmt.Sprintf("send pcm data finished, text: [%s]", text), logTag)
		}
	}()

	rte.OnStartDone()
}

// OnCmd receives cmd from rte graph.
// current supported cmd:
//   - name: flush
//     example:
//     {"name": "flush"}
func (e *elevenlabsTTSExtension) OnCmd(
	rte rtego.Rte,
	cmd rtego.Cmd,
) {
	cmdName, err := cmd.CmdName()
	if err != nil {
		slog.Error(fmt.Sprintf("OnCmd get name failed, err: %v", err), logTag)
		rte.ReturnString(rtego.Error, "error", cmd)
		return
	}

	slog.Info(fmt.Sprintf("OnCmd %s", cmdInFlush), logTag)

	switch cmdName {
	case cmdInFlush:
		// clear chan
		// TODO
		// textChan = make(chan string, cap(textChan))

		// send out
		outCmd, err := rtego.NewCmd(cmdOutFlush)
		if err != nil {
			slog.Error(fmt.Sprintf("new cmd %s failed, err: %v", cmdOutFlush, err), logTag)
			rte.ReturnString(rtego.Error, "error", cmd)
			return
		}

		if err := rte.SendCmd(outCmd, nil); err != nil {
			slog.Error(fmt.Sprintf("send cmd %s failed, err: %v", cmdOutFlush, err), logTag)
			rte.ReturnString(rtego.Error, "error", cmd)
			return
		} else {
			slog.Info(fmt.Sprintf("cmd %s sent", cmdOutFlush), logTag)
		}
	}

	rte.ReturnString(rtego.Ok, "ok", cmd)
}

// OnData receives data from rte graph.
// current supported data:
//   - name: text_data
//     example:
//     {name: text_data, properties: {text: "hello"}
func (e *elevenlabsTTSExtension) OnData(
	rte rtego.Rte,
	data rtego.Data,
) {
	text, err := data.GetPropertyString(dataInTextDataPropertyText)
	if err != nil {
		slog.Warn(fmt.Sprintf("OnData GetProperty %s failed, err: %v", dataInTextDataPropertyText, err), logTag)
		return
	}

	if len(text) == 0 {
		slog.Debug("OnData text is empty, ignored", logTag)
		return
	}

	slog.Info(fmt.Sprintf("OnData input text: [%s]", text), logTag)

	go func() {
		textChan <- text
	}()
}

func init() {
	slog.Info("elevenlabs_tts extension init", logTag)

	// Register addon
	rtego.RegisterAddonAsExtension(
		"elevenlabs_tts",
		rtego.NewDefaultExtensionAddon(newElevenlabsTTSExtension),
	)
}

File: agents/addon/extension/elevenlabs_tts_python/elevenlabs_tts_extension.py
#
#
# Agora Real Time Engagement
# Created by XinHui Li in 2024-07.
# Copyright (c) 2024 Agora IO. All rights reserved.
#
#

import logging
import queue
import threading
import time

from rte_runtime_python import (
    Addon,
    Extension,
    register_addon_as_extension,
    Rte,
    Cmd,
    CmdResult,
    StatusCode,
    Data,
    MetadataInfo,
)
from elevenlabs_tts import default_elevenlabs_tts_config, ElevenlabsTTS
from pcm import PcmConfig, Pcm

CMD_IN_FLUSH = "flush"
CMD_OUT_FLUSH = "flush"

DATA_IN_TEXT_DATA_PROPERTY_TEXT = "text"

PROPERTY_API_KEY = "api_key"  # Required
PROPERTY_MODEL_ID = "model_id"  # Optional
PROPERTY_OPTIMIZE_STREAMING_LATENCY = "optimize_streaming_latency"  # Optional
PROPERTY_REQUEST_TIMEOUT_SECONDS = "request_timeout_seconds"  # Optional
PROPERTY_SIMILARITY_BOOST = "similarity_boost"  # Optional
PROPERTY_SPEAKER_BOOST = "speaker_boost"  # Optional
PROPERTY_STABILITY = "stability"  # Optional
PROPERTY_STYLE = "style"  # Optional

logging.basicConfig(
    format="%(asctime)s - %(name)s - %(levelname)s - %(process)d - [%(filename)s:%(lineno)d] - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    level=logging.INFO,
    encoding="utf-8",
)


class Message:
    def __init__(self, text: str, received_ts: int) -> None:
        self.text = text
        self.received_ts = received_ts


class ElevenlabsTTSExtension(Extension):
    def on_init(self, rte: Rte, manifest: MetadataInfo, property: MetadataInfo) -> None:
        logging.info("on_init")

        self.elevenlabs_tts = None
        self.outdate_ts = 0
        self.pcm = None
        self.pcm_frame_size = 0
        self.text_queue = queue.Queue(maxsize=1024)

        rte.on_init_done(manifest, property)

    def on_start(self, rte: Rte) -> None:
        logging.info("on_start")

        # prepare configuration
        elevenlabs_tts_config = default_elevenlabs_tts_config()

        try:
            elevenlabs_tts_config.api_key = rte.get_property_string(PROPERTY_API_KEY)
        except Exception as e:
            logging.warning(
                f"on_start get_property_string {PROPERTY_API_KEY} error: {e}"
            )
            return

        try:
            model_id = rte.get_property_string(PROPERTY_MODEL_ID)
            if len(model_id) > 0:
                elevenlabs_tts_config.model_id = model_id
        except Exception as e:
            logging.warning(
                f"on_start get_property_string {PROPERTY_MODEL_ID} error: {e}"
            )

        try:
            optimize_streaming_latency = rte.get_property_int(
                PROPERTY_OPTIMIZE_STREAMING_LATENCY
            )
            if optimize_streaming_latency > 0:
                elevenlabs_tts_config.optimize_streaming_latency = (
                    optimize_streaming_latency
                )
        except Exception as e:
            logging.warning(
                f"on_start get_property_int {PROPERTY_OPTIMIZE_STREAMING_LATENCY} error: {e}"
            )

        try:
            request_timeout_seconds = rte.get_property_int(
                PROPERTY_REQUEST_TIMEOUT_SECONDS
            )
            if request_timeout_seconds > 0:
                elevenlabs_tts_config.request_timeout_seconds = request_timeout_seconds
        except Exception as e:
            logging.warning(
                f"on_start get_property_int {PROPERTY_REQUEST_TIMEOUT_SECONDS} error: {e}"
            )

        try:
            elevenlabs_tts_config.similarity_boost = rte.get_property_float(
                PROPERTY_SIMILARITY_BOOST
            )
        except Exception as e:
            logging.warning(
                f"on_start get_property_float {PROPERTY_SIMILARITY_BOOST} error: {e}"
            )

        try:
            elevenlabs_tts_config.speaker_boost = rte.get_property_bool(
                PROPERTY_SPEAKER_BOOST
            )
        except Exception as e:
            logging.warning(
                f"on_start get_property_bool {PROPERTY_SPEAKER_BOOST} error: {e}"
            )

        try:
            elevenlabs_tts_config.stability = rte.get_property_float(PROPERTY_STABILITY)
        except Exception as e:
            logging.warning(
                f"on_start get_property_float {PROPERTY_STABILITY} error: {e}"
            )

        try:
            elevenlabs_tts_config.style = rte.get_property_float(PROPERTY_STYLE)
        except Exception as e:
            logging.warning(f"on_start get_property_float {PROPERTY_STYLE} error: {e}")

        # create elevenlabsTTS instance
        self.elevenlabs_tts = ElevenlabsTTS(elevenlabs_tts_config)

        logging.info(
            f"ElevenlabsTTS succeed with model_id: {self.elevenlabs_tts.config.model_id}, VoiceId: {self.elevenlabs_tts.config.voice_id}"
        )

        # create pcm instance
        self.pcm = Pcm(PcmConfig())
        self.pcm_frame_size = self.pcm.get_pcm_frame_size()

        threading.Thread(target=self.process_text_queue, args=(rte,)).start()

        rte.on_start_done()

    def on_stop(self, rte: Rte) -> None:
        logging.info("on_stop")
        rte.on_stop_done()

    def on_deinit(self, rte: Rte) -> None:
        logging.info("on_deinit")
        rte.on_deinit_done()

    def on_cmd(self, rte: Rte, cmd: Cmd) -> None:
        """
        on_cmd receives cmd from rte graph.
        current supported cmd:
          - name: flush
            example:
            {"name": "flush"}
        """
        logging.info("on_cmd")
        cmd_name = cmd.get_name()

        logging.info(f"on_cmd [{cmd_name}]")

        if cmd_name is CMD_IN_FLUSH:
            self.outdate_ts = int(time.time() * 1000000)

            # send out
            out_cmd = cmd.create(CMD_OUT_FLUSH)
            rte.send_cmd(out_cmd)

        cmd_result = CmdResult.create(StatusCode.OK)
        cmd_result.set_property_string("detail", "success")
        rte.return_result(cmd_result, cmd)

    def on_data(self, rte: Rte, data: Data) -> None:
        """
        on_data receives data from rte graph.
        current supported data:
          - name: text_data
            example:
            {name: text_data, properties: {text: "hello"}
        """
        logging.info("on_data")

        try:
            text = data.get_property_string(DATA_IN_TEXT_DATA_PROPERTY_TEXT)
        except Exception as e:
            logging.warning(
                f"on_data get_property_string {DATA_IN_TEXT_DATA_PROPERTY_TEXT} error: {e}"
            )
            return

        if len(text) == 0:
            logging.debug("on_data text is empty, ignored")
            return

        logging.info(f"OnData input text: [{text}]")

        self.text_queue.put(Message(text, int(time.time() * 1000000)))

    def process_text_queue(self, rte: Rte):
        logging.info("process_text_queue")

        while True:
            msg = self.text_queue.get()
            if msg.received_ts < self.outdate_ts:
                logging.info(
                    f"textChan interrupt and flushing for input text: [{msg.text}], received_ts: {msg.received_ts}, outdate_ts: {self.outdate_ts}"
                )
                continue

            start_time = time.time()
            buf = self.pcm.new_buf()
            first_frame_latency = 0
            n = 0
            pcm_frame_read = 0
            read_bytes = 0
            sent_frames = 0

            audio_stream = self.elevenlabs_tts.text_to_speech_stream(msg)

            for chunk in self.pcm.read_pcm_stream(audio_stream, self.pcm_frame_size):
                if msg.received_ts < self.outdate_ts:
                    logging.info(
                        f"textChan interrupt and flushing for input text: [{msg.text}], received_ts: {msg.received_ts}, outdate_ts: {self.outdate_ts}"
                    )
                    break

                if not chunk:
                    logging.info("read pcm stream EOF")
                    break

                n = len(chunk)
                read_bytes += n
                pcm_frame_read += n

                if pcm_frame_read != self.pcm.get_pcm_frame_size():
                    logging.debug(
                        f"the number of bytes read is [{pcm_frame_read}] inconsistent with pcm frame size",
                    )
                    continue

                self.pcm.send(rte, buf)
                buf = self.pcm.new_buf()
                pcm_frame_read = 0
                sent_frames += 1

                if first_frame_latency == 0:
                    first_frame_latency = int((time.time() - start_time) * 1000)
                    logging.info(
                        f"first frame available for text: [{msg.text}], received_ts: {msg.received_ts}, first_frame_latency: {first_frame_latency}ms",
                    )

                logging.debug(f"sending pcm data, text: [{msg.text}]")

            if pcm_frame_read > 0:
                self.pcm.send(rte, buf)
                sent_frames += 1
                logging.info(
                    f"sending pcm remain data, text: [{msg.text}], pcm_frame_read: {pcm_frame_read}"
                )

            finish_latency = int((time.time() - start_time) * 1000)
            logging.info(
                f"send pcm data finished, text: [{msg.text}], received_ts: {msg.received_ts}, read_bytes: {read_bytes}, sent_frames: {sent_frames}, \
                first_frame_latency: {first_frame_latency}ms, finish_latency: {finish_latency}ms"
            )


@register_addon_as_extension("elevenlabs_tts_python")
class ElevenlabsTTSExtensionAddon(Addon):
    def on_init(self, rte: Rte, manifest, property) -> None:
        logging.info("on_init")
        rte.on_init_done(manifest, property)
        return

    def on_create_instance(self, rte: Rte, addon_name: str) -> Extension:
        logging.info("on_create_instance")
        return ElevenlabsTTSExtension(addon_name)

    def on_deinit(self, rte: Rte) -> None:
        logging.info("on_deinit")
        rte.on_deinit_done()
        return
